### PR TITLE
Add pretty printing for splat(f)

### PR DESCRIPTION
**Pull request by Seelengrab**
When debugging calls with different invocations of `splat`, this helps to identify which function was being splatted into.

master:

```julia
julia> Base.splat(cos)
#92 (generic function with 1 method)

julia> f = Base.splat(sin)
#92 (generic function with 1 method)

julia> f(1)
0.8414709848078965

julia> [f]
1-element Vector{Base.var"#92#93"{typeof(sin)}}:
 #92 (generic function with 1 method)

julia> using Logging

julia> @info f
[ Info: #92
```

This PR:

```julia
julia> Base.splat(cos)
splat(cos)

julia> f = Base.splat(sin)
splat(sin)

julia> f(1)
0.8414709848078965

julia> [f]
1-element Vector{Base.Splat{typeof(sin)}}:
 splat(sin)

julia> @info f
[ Info: splat(sin)
```

Should `splat` be exported? It's pretty useful.

### CODE REVIEW

**Comment by Seelengrab** [comment]
Well it'd be lying to you, as that's not how its implemented :shrug: That used to be the old implementation though. Changing it was necessary to allow for easier printing. 

Compared to the old way of just printing whatever number anonymous function the result of `splat` happened to be, this representation is already a big step up and allows you to do `?Base.splat(sin)` (we'd have to export it to make `?splat` work) and you'll get the docs for it, where it says that this is equivalent to the lambda version.

Since it's not exported, I don't think it's used very wildly right now. I'm working on a package ([PropCheck.jl](https://github.com/Seelengrab/PropCheck.jl)) where using `splat` to interact with it makes everything a bunch easier (see e.g. the tests in [BigNums.jl](https://github.com/Seelengrab/BigNums.jl/blob/master/test/runtests.jl)). I got annoyed with not knowing which function was wrapped, hence this PR.

**Comment by Seelengrab** [comment]
I've incorporated the feedback, fixed printing, added the deprecation and replaced uses of the old syntax in the rest of Base. It's also rebased to the current master. Now let's hope nothing broke :grimacing:

File: base/operators.jl [hunk 1]
```diff
@@ -1186,27 +1186,38 @@ used to implement specialized methods.
...
+Splat(+)
+
+julia> my_add((1,2,3))
+6
 ```
 """
-splat(f) = args->f(args...)
+struct Splat{F} <: Function
+    f::F
+end
```
Comment (vtjnash) [inline base/operators.jl:1217]:
```suggestion
    Splat(f) = new{Core.Typeof(f)}(f)
end
```

**Comment by vtjnash** [comment]
Add export?

**Comment by Seelengrab** [comment]
> $ /etc/buildkite-agent/plugins/github-com-JuliaCI-external-buildkite-buildkite-plugin-v1/hooks/post-checkout
> \--
>   | fatal: destination path '.buildkite' already exists and is not an empty directory.
>   | 🚨 Error: The plugin external-buildkite post-checkout hook exited with status 128

Uuh.. Unrelated? Doesn't look good though..

---

In either case, I've also added a NEWS.md entry since CI was green last night. Should be good to go now.

**Comment by Seelengrab** [comment]
The macos failure seems to stem from `Libgit2`:

> Test Failed at /Users/julia/buildbot/worker-tabularasa/tester_macos64/build/share/julia/stdlib/v1.9/LibGit2/test/libgit2.jl:1754
>  Expression: LibGit2.headname(repo) == master_branch
>   Evaluated: "main" == "master"

**Comment by Seelengrab** [comment]
I don't know where this should fall in the release cycle, but it'd be nice to get this in a future 1.9 :)

**Comment by Seelengrab** [comment]
Rebase on `master`!

**Comment by Seelengrab** [comment]
Failure in BuildKite macOS is `cmdlineargs`, seems unrelated.

**Pull request merged**